src/dtx/dtx_rpc.c

```diff
@@ -670,11 +700,18 @@ dtx_rpc_prep(struct ds_cont_child *cont,d_list_t *dti_list,  struct dtx_entry **
 	}
 
 	/* Use helper ULT to handle DTX RPC if there are enough helper XS. */
-	if (dss_has_enough_helper())
-		rc = dss_ult_create(dtx_rpc_helper, dca, DSS_XS_IOFW, dca->dca_tgtid, 0,
-				    &dca->dca_helper);
-	else
-		rc = dtx_rpc_internal(dca);
+	if (dss_has_enough_helper()) {
+		rc = ABT_eventual_create(0, &dca->dca_chore_eventual);
```

Comment:
Why non-coll case need the eventual, but coll case does not?

Reply:
Is it possible to not create the dca_chore_eventual, instead, depends on the dra->dra_future to check whether related operation has been done or not, then non-coll case and coll case can share the same mechanism or logic.

Reply:
> Why non-coll case need the eventual, but coll case does not?

@Nasf-Fan, the `dca_chore_eventual` is for `dtx_rpc_post` to know when it's safe to access `dra_future`.

> Is it possible to not create the dca_chore_eventual, instead, depends on the dra->dra_future to check whether related operation has been done or not, then non-coll case and coll case can share the same mechanism or logic.

AFAICS, it's possible, though not easy. The main trouble is that the number of slots in the future is computed _by_ the chore, so we can't create the future until the chore has run to the point where that number is computed. Also, even if we really want to make such a change in this PR, it will take more changes to share the mechanism with `dtx_coll_rpc_helper`, since the two code paths use different data structures.

So, what do you think?

Reply:
> > Why non-coll case need the eventual, but coll case does not?
> 
> @Nasf-Fan, the `dca_chore_eventual` is for `dtx_rpc_post` to know when it's safe to access `dra_future`.
> 
> > Is it possible to not create the dca_chore_eventual, instead, depends on the dra->dra_future to check whether related operation has been done or not, then non-coll case and coll case can share the same mechanism or logic.
> 
> AFAICS, it's possible, though not easy. The main trouble is that the number of slots in the future is computed _by_ the chore, so we can't create the future until the chore has run to the point where that number is computed. Also, even if we really want to make such a change in this PR, it will take more changes to share the mechanism with `dtx_coll_rpc_helper`, since the two code paths use different data structures.
> 
> So, what do you think?

Then ignore my former comment and go ahead.